### PR TITLE
[Feature] Binlog connector can read binlog from tablet

### PR DESCRIPTION
**Pull request by banmoy**
## What type of PR is this：
- [ ] BugFix
- [X] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Binlog connector can read binlog from tablet using `BinlogReader`, and there are two use cases
1. scan **full of binlog** using `select * from t [_BINLOG_]` which is only used for testing currently
2. realtime mv read binlog from **the given offset** for incremental computation

Binlog connector will distinguish the two cases using the flag `is_stream_pipeline`,  `false` for case 1, and `true` for case 2. This PR includes codes to support both cases, but actually the code can only run in case 1 currently, and there are still some problems for case 2 which need more work to debug the end-to-end mv pipeline. I still include the code for case 2 in this PR so that reviewers can have a full context.

## How to use
You can test case 1 in following steps

1. create table with binlog enabled
```
CREATE TABLE `test_binlog` (
  `c0` int(11) NULL COMMENT "",
  `c1` int(11) NULL COMMENT "",
  `c2` varchar(100) NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`c0`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`c0`) BUCKETS 2
PROPERTIES (
    "replication_num" = "1",
    "binlog_enable" = "true",
    "binlog_ttl_second" = "3600",
    "binlog_max_size" = "107374182400"
);
```
2.  insert some data to the table

```
INSERT INTO `test_binlog` VALUES (1, 2, "12"), (2, 3, "23"), (3, 4, "34"), (4, 5, "45");
```

3. enable mv planner and execute select with hint `[_BINLOG_]`
```
SET enable_mv_planner=true;
SELECT * from `test_binlog` [_BINLOG_];
```

And the result will be like this
```
+------+------+------+------------+-----------------+----------------+-------------------+
| c0   | c1   | c2   | _binlog_op | _binlog_version | _binlog_seq_id | _binlog_timestamp |
+------+------+------+------------+-----------------+----------------+-------------------+
|    3 |    4 | 34   |          0 |               2 |              0 |  1677571530000000 |
|    1 |    2 | 12   |          0 |               2 |              0 |  1677571530000000 |
|    2 |    3 | 23   |          0 |               2 |              1 |  1677571530000000 |
|    4 |    5 | 45   |          0 |               2 |              2 |  1677571530000000 |
+------+------+------+------------+-----------------+----------------+-------------------+
4 rows in set (0.02 sec)

```

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Review by decster — APPROVED** [review]

**Comment by banmoy** [comment]
run starrocks_admit_test

File: be/src/connector/binlog_connector.h [hunk 1]
```diff
@@ -87,13 +83,23 @@ class BinlogDataSource final : public StreamDataSource {
...
     RuntimeState* _runtime_state = nullptr;
+    bool _is_stream_pipeline = false;
+
     TabletSharedPtr _tablet;
-    // TODO this will be used by BinlogReader
     Schema _binlog_read_schema;
+    BinlogReaderSharedPtr _binlog_reader;
+
+    // whether to need do a seek before read data
+    std::atomic<bool> _need_seek_binlog{true};
```
Comment (smartlxh) [inline be/src/connector/binlog_connector.h:99]: Under what situation will there be concurrent access?

Comment (banmoy) [reply be/src/connector/binlog_connector.h:99]: `set_offset()` and `get_next()` may be in different threads?

**Review by murphyatwork — APPROVED** [review]

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**